File: exercises/dotnet/PaymentService/Domain/Identifiers/PaymentContextKey.cs
using SharedObservability;

namespace PaymentService.Domain.Identifiers;

/// <summary>
/// Payment's typed identifier vocabulary.
/// <see cref="AuthorizationId"/> is purely Payment's; the others use wire
/// strings agreed with Order (order.id, customer.id).
/// </summary>
public sealed class PaymentContextKey
{
    public static readonly PaymentContextKey OrderId = new("order.id");
    public static readonly PaymentContextKey CustomerId = new("customer.id");
    public static readonly PaymentContextKey AuthorizationId = new("authorization.id");

    public string Key { get; }

    private PaymentContextKey(string key) => Key = key;

    /// <summary>
    /// Bind this key to a concrete value, returning an <see cref="IDomainIdentifier"/>.
    /// </summary>
    public IDomainIdentifier Of(string value) => new BoundIdentifier(Key, value);

    private sealed record BoundIdentifier(string Key, string Value) : IDomainIdentifier;
}

File: exercises/quarkus/payment-service/src/main/java/com/example/payment/domain/identifier/PaymentContextKey.java
package com.example.payment.domain.identifier;

import com.example.workshop.observability.DomainIdentifier;

import java.util.Objects;

/**
 * Payment's typed identifier vocabulary.
 *
 * <p>{@link #AUTHORIZATION_ID} is purely Payment's; the others use wire
 * strings agreed with Order ({@code order.id}, {@code customer.id}).
 */
public enum PaymentContextKey {

    ORDER_ID("order.id"),
    CUSTOMER_ID("customer.id"),
    AUTHORIZATION_ID("authorization.id");

    private final String key;

    PaymentContextKey(String key) {
        this.key = key;
    }

    public String key() {
        return key;
    }

    public DomainIdentifier of(String value) {
        Objects.requireNonNull(value, () -> name() + " value");
        return new SimpleIdentifier(key, value);
    }

    private record SimpleIdentifier(String key, String value)
            implements DomainIdentifier {
    }
}
